xwiki-platform-core/xwiki-platform-component/xwiki-platform-component-wiki/src/main/java/org/xwiki/component/wiki/internal/bridge/DefaultWikiObjectComponentManagerEventListener.java

```diff
@@ -0,0 +1,164 @@
+import org.xwiki.component.manager.ComponentManager;
+import org.xwiki.component.wiki.WikiObjectComponentBuilder;
+import org.xwiki.component.wiki.internal.DefaultWikiComponentManagerEventListener;
+import org.xwiki.model.reference.DocumentReference;
+import org.xwiki.model.reference.EntityReferenceSerializer;
+import org.xwiki.observation.AbstractEventListener;
+import org.xwiki.observation.EventListener;
+import org.xwiki.observation.event.Event;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.internal.event.XObjectAddedEvent;
+import com.xpn.xwiki.internal.event.XObjectDeletedEvent;
+import com.xpn.xwiki.internal.event.XObjectEvent;
+import com.xpn.xwiki.internal.event.XObjectUpdatedEvent;
+import com.xpn.xwiki.objects.BaseObjectReference;
+
+/**
+ * This {@link EventListener} is responsible for registering wiki components derived from XObjects using the same
+ * principle as {@link DefaultWikiComponentManagerEventListener}.
+ * 
+ * @version $Id$
+ * @since 9.5RC1
+ */
+@Component
+@Named(DefaultWikiObjectComponentManagerEventListener.EVENT_LISTENER_NAME)
+@Singleton
+public class DefaultWikiObjectComponentManagerEventListener extends AbstractEventListener
+{
+    /**
+     * The event listener name.
+     */
+    public static final String EVENT_LISTENER_NAME = "defaultWikiObjectComponentManagerEventListener";
+
+    @Inject
+    private ComponentManager componentManager;
+
+    @Inject
+    private WikiObjectComponentManagerEventListenerProxy wikiObjectComponentManagerEventListenerProxy;
+
+    @Inject
+    private EntityReferenceSerializer<String> defaultEntityReferenceSerializer;
+
+    @Inject
+    @Named("local")
+    private EntityReferenceSerializer<String> localEntityReferenceSerializer;
+
+    @Inject
+    private Logger logger;
+
+    @Inject
+    private Provider<XWikiContext> xcontextProvider;
+
+    /**
+     * Builds a new {@link DefaultWikiObjectComponentManagerEventListener}.
+     */
+    public DefaultWikiObjectComponentManagerEventListener()
+    {
+        super(DefaultWikiObjectComponentManagerEventListener.EVENT_LISTENER_NAME,
+                new ApplicationReadyEvent(), new WikiReadyEvent(),
+                new XObjectAddedEvent(), new XObjectUpdatedEvent(), new XObjectDeletedEvent());
+    }
+
+    @Override
+    public void onEvent(Event event, Object source, Object data)
+    {
+        /* If we are dealing with an XObject related event, then it might be related to a
+         * WikiObjectComponentBuilder */
+        if (event instanceof XObjectEvent) {
+            XWikiDocument document = (XWikiDocument) source;
+            // If we are dealing with a XObjectDeletedEvent, we have to retrieve the original XObject document.
+            if (event instanceof XObjectDeletedEvent) {
+                document = document.getOriginalDocument();
+            }
+            BaseObjectReference baseObjectReference = document.getXObject().getReference();
```

Comment:
Document.getXObject() return the first object of the document which is really not what you want (unless you are lucky).

You should pass the reference given by the event to getXObject().